hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java

```diff
@@ -457,6 +459,7 @@ public abstract class AbstractEntityPersister
 	protected final BasicEntityPropertyMapping propertyMapping;
 
 	private final boolean implementsLifecycle;
+	private volatile List<UniqueKeyEntry> uniqueKeyEntries = null;//lazily initialized
```

Comment:
Ok maybe it's better to initialize it in `doLateInit`, still I don't see why that would allow me to avod the `volatile` - the way such fields get initialized after the constructor always makes me a bit uncomfortable.

Reply:
I was uncomfortable with that as well, but I think you were the one who told me that this is fine because the initialization phase is single threaded and the whole thing is finally safely published ;)

Reply:
Yes that's why it seems to work in practice :)
I'd still prefer to improve the design though.. not overnight and not urgently, but I tend to try leave it better whenever I patch it.

Let's put the "volatile" issue to the side for a moment - one reason I had it lazily initialized was that I assumed that perhaps not all persister's `List<UniqueKeyEntry>` is useful? So in interest of saving some memory - but you'd know better, if you suggest this is unlikely it might as well be made eager.

Reply:
Whenever an entity is initialized the `EntityUniqueKey` for the unique properties of the entity type are registered, so this will essentially always be needed, so eager is the way to go IMO.

Reply:
ok, thanks!

Reply:
It seems that if I don't make this initialization lazy, it uncovers an existing bug. I'll keep it lazy for now so we can focus on one thing at a time..

Reply:
Can you post a link to a JIRA for that bug here please?